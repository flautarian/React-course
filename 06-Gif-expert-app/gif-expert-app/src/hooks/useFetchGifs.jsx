import { useEffect, useState } from "react";
import { fillGifs } from "../helpers/get-gifs";

export const UseFetchGifs = ( category ) => {

    const [images, setImages] = useState([]);
    const [isLoading, setIsLoading] = useState( true );

    const getImages = async() => {
        const newImages = await fillGifs( category );
        setImages(newImages);
        setIsLoading(false);
    }

    useEffect(() => {
        getImages();
    }, []) 

    return {
        images,
        isLoading
    }
}
