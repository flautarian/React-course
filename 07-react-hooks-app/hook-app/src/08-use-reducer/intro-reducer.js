const initialState = [{
    id: 1,
    todo: 'Collect soul stone',
    done: false,
}];


const todoReducer = (state = initialState, action) => {
    if(!!action && action.type === "[TODO] add todo"){
        return [...state, action.payload];
    }
    return state;
}

const newTodo = {
    id: 2,
    todo: 'Collect power stone',
    done: false,
}

const addTodoAction = {
    type: "[TODO] add todo",
    payload: newTodo,
}

let todos = todoReducer();

todos = todoReducer(todos, addTodoAction);

console.log(todos);