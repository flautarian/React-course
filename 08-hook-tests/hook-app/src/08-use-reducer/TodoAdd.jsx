import React, { Fragment, useEffect, useState } from 'react'

export const TodoAdd = ({ createTodo }) => {

    const [form, setForm] = useState({
        desc: "",
        done: false
    });

    const { desc, done } = form;

    const onFormChange = ({ target }) => {
        const { name, value, checked, type } = target;
        setForm({
            ...form,
            [name]: type === "checkbox" ? checked : value
        })
    }

    useEffect(() => { }, [form]);

    const onNewTodo = (event) => {
        event.preventDefault();
        const todo = {
            id: new Date().getTime(),
            desc,
            done,
        }
        createTodo(todo);
    }

    return (
        <Fragment>
            <form onSubmit={(event) => onNewTodo(event)}>
                <input
                    type="text"
                    placeholder='What to do?'
                    className='form-control'
                    value={desc}
                    name='desc'
                    onChange={onFormChange} ></input>
                Done: 
                <input className='mt-2' type="checkbox" name='done' value={done} onChange={onFormChange} ></input>
                <br/>
                <button type='submit' className='btn btn-outline-primary mt-2'>Add Todo</button>
            </form>
        </Fragment>
    )
}
