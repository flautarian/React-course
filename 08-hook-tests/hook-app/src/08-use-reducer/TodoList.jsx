import React, { useEffect } from 'react'
import { TodoItem } from './TodoItem'

export const TodoList = ({ list, toDelete, toToggle }) => {

    useEffect(() => { }, [list]);

    return (
        <ul className='list-group'>
            {
                list.map(todo => (
                    <TodoItem key={todo.id} item={todo} toDelete={toDelete} toToggle={toToggle} />
                ))
            }
        </ul>
    )
}
