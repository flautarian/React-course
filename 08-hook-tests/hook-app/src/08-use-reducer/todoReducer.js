

export const todoReducer = (initialState, action) => {
    switch (action.type) {
        case 'add':
            return [...initialState, action.payload];
        case 'remove':
            return initialState.filter(element => element.id !== action.id);
        case 'toggleDone':
            return initialState.map(element => {
                if (element.id === action.id) element.done = !element.done;
                return element;
            })
        default:
            return initialState;
    }
}