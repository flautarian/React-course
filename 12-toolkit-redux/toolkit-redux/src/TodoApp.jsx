import React from 'react'
import { useState } from 'react';
import { Fragment } from 'react'
import { useGetTodoQuery, useGetTodosQuery } from './store/apis';

export const TodoApp = () => {


    const [todoId, setTodoId] = useState(1);

    const { data: todo, isLoading } = useGetTodoQuery( todoId );

    return (
        <Fragment>
            <h1> Todos - RTK Query</h1>

            <hr />

            {isLoading && <h4>Is Loading...</h4>}

            <pre>{ JSON.stringify( todo ) }</pre>

            <button disabled={isLoading} onClick={() => setTodoId(todoId > 1 ? todoId - 1 : 0)}>Prev todo</button>

            <button disabled={isLoading} onClick={() => setTodoId(todoId + 1)}>Next todo</button>
            {/* <ul>
                {
                    todos.map(todo => (
                        <li key={todo.id}>
                            <strong>{todo.completed ? 'Done' : 'Pending'}</strong> - {todo.title}
                        </li>
                    ))
                }
            </ul> */}


        </Fragment>
    )
}
