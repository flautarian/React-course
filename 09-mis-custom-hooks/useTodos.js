import React, { useEffect, useReducer } from 'react';
import { todoReducer } from '../08-use-reducer/todoReducer';


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const useTodos = ({ initialState }) => {

    const [state, dispatch] = useReducer(todoReducer, initialState, init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(state));
    }, [state])

    const deleteTodo = (id) => {
        let todoAction = {
            type: "remove",
            id,
        }
        dispatch(todoAction);
    }

    const toggleTodo = (id) => {
        let todoAction = {
            type: "toggleDone",
            id,
        }
        dispatch(todoAction);
    }

    const createTodo = (todo) => {
        let todoAction = {
            type: "add",
            payload: todo,
        }
        dispatch(todoAction);
    }

    return {
        state, 
        todosCount : state.length, 
        todosRemaining: state.filter(todo => !todo.done).length,
        deleteTodo,
        toggleTodo,
        createTodo
    }
}
