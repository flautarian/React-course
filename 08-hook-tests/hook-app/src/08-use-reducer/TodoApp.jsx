import React, { Fragment, useEffect, useReducer } from 'react'
import { useTodos } from '../hooks/useTodos';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer'

export const TodoApp = () => {

    const { state, todosCount, todosRemaining, deleteTodo, toggleTodo, createTodo } = useTodos([]);

    return (
        <Fragment>
            <h2>TodoApp ({todosCount})
                <small>
                    Pendent: ({todosRemaining})
                </small>
            </h2>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList list={state} toDelete={deleteTodo} toToggle={toggleTodo} />
                </div>

                <div className="col-5">
                    <h4>Add todo Action:</h4>
                    <hr />
                    <TodoAdd createTodo={createTodo} />
                </div>
            </div>

        </Fragment>
    )
}
