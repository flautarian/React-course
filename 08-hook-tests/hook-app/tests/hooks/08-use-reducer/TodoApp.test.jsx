import { render, screen } from '@testing-library/react';
import { TodoApp } from '../../../src/08-use-reducer/TodoApp';
import { useTodos } from '../../../src/hooks/useTodos';

jest.mock('../../../src/hooks/useTodos');

describe('TodoApp test', () => {

    useTodos.mockReturnValue({
        state: [{ id: 1, desc: "desc1", done: false },
        { id: 2, desc: "desc2", done: true }],
        todosCount: 2,
        todosRemaining: 1,
        deleteTodo: jest.fn(),
        toggleTodo: jest.fn(),
        createTodo: jest.fn(),
    });

    test('should match snapshot', () => {
        const { container } = render(<TodoApp />);
        expect(container).toMatchSnapshot();
    });

    test('asdf', () => {
        const { container } = render(<TodoApp />);
        expect(screen.getByText('desc1')).toBeTruthy();
        expect(screen.getByText('desc2')).toBeTruthy();
        expect(screen.getByRole('textbox')).toBeTruthy();
    });
})