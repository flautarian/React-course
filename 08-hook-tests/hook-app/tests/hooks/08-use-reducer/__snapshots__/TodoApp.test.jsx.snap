// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp test should match snapshot 1`] = `
<div>
  <h2>
    TodoApp (
    2
    )
    <small>
      Pendent: (
      1
      )
    </small>
  </h2>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center"
          >
            desc1
          </span>
          <div>
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
            <button
              class="btn btn-warning"
            >
              To do
            </button>
          </div>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center"
          >
            desc2
          </span>
          <div>
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
            <button
              class="btn btn-success"
            >
              Done
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add todo Action:
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="desc"
          placeholder="What to do?"
          type="text"
          value=""
        />
        Done:
        <input
          class="mt-2"
          name="done"
          type="checkbox"
          value="false"
        />
        <br />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Add Todo
        </button>
      </form>
    </div>
  </div>
</div>
`;
